python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -1158,6 +1160,176 @@ def f():
+      .extracting(PythonType::name)
+      .isEqualTo("int");
+  }
+
+  @Test
+  void inferLoopOverLiteralVarTypeCallable() {
+    var root = inferTypes("""
+      def foo():
+        print("foo")
+        
+      def bar():
+        print("bar")
+        
+      def f():
+        l = [foo, bar]
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(2), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnionType.class);
+    var candidates = ((UnionType) iType).candidates();
+    Assertions.assertThat(candidates)
+      .allMatch(FunctionType.class::isInstance)
+      .extracting(PythonType::name)
+      .contains("foo", "bar");
+  }
+
+
+  @Test
+  void inferLoopOverUnknownVarType() {
+    var root = inferTypes("""
+      def f(l):
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(0), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnknownType.class)
+      .isSameAs(PythonType.UNKNOWN);
+  }
+
+  @Test
+  void inferLoopOverNonIterableVarType() {
+    var root = inferTypes("""
+      def f():
+        l = 1
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(0), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnknownType.class)
+      .isSameAs(PythonType.UNKNOWN);
+  }
+
+  @Test
+  void inferLoopOverNonIterableVarTypeInTry() {
+    var root = inferTypes("""
+      def f():
+        try:
+          l = 1
+          for i in l:
+            i
+        except:
+          ...
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(0), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnknownType.class)
+      .isSameAs(PythonType.UNKNOWN);
+  }
+
+  @Test
+  void hasUnknownListItemType() {
+    var root = inferTypes("""
+      def f(i):
+        l = [1,2,l]
```

Comment:
Did you mean for the last item to be `i` or `l`? Both could make sense but as I read the test I have the feeling you were going for `i` here.